Set canvas size based on text length

behackl opened the pull request:
## List of Changes
Rendering long strings as `Text` lets the text be outside of the predefined canvas, leading to only a part of the string being converted to the svg. This PR sets the canvas width dependent on the text width.

## Motivation

Resolves #469.

## Testing Status

The issue is fixed, but the test I wanted to add is currently not working.

Depends on #492 


## Acknowledgement
- [x] I have read the [Contributing Guidelines](https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP))

behackl commented:
The tests here fail (at least locally, so I'd be surprised if they would pass in the pipeline), because it seems it is not possible to call `Text` within a test. Is there a workaround I could use, @ManimCommunity/tests ?

naveen521kk commented on tests/test_text.py:
```diff
@@ -0,0 +1,9 @@
+    is resolved."""
+
+    t = Text("a" * 150, font="Arial")
+    assert len(t.text) == len(t.submobjects)
```
Pretty much this won't work at all. Because I had found it out while working on #471 that Cairo will produce some empty `path` in SVG. Manim just ignores those empty paths. So checking this won't work.
https://github.com/ManimCommunity/manim/blob/650d1528fbabf9e326433e7e702e6233c6f9afeb/manim/mobject/svg/svg_mobject.py#L90-L93

behackl replied:
The fact that the text length does not coincide with the number of submobjects is precisely what causes the error in #469, you can check that we have this equality for `Text("a" * 100)`, but not for `Text("a" * 150)`.

naveen521kk replied:
Pretty much I tried a lot on that. and possibly removing
https://github.com/ManimCommunity/manim/blob/master/manim/mobject/svg/text_mobject.py#L152-L173
worked out. With some of its references.

naveen521kk replied:
That worked out for Pango but I don't think it will work here sadly.

behackl replied:
I am not sure what you are saying. What do you mean by "removing" `gen_chars`?

Just to be clear: the changes proposed here simply guarantee that the entire text will fit onto the canvas the svg is rendered on, ensuring that arbitrarily long strings can be handled. (The canvas actually gets way to large by this approach, but I could not find a significant impact to render speed.)

In particular, these changes allow the string given in #469 and the test string `Text("a" * 120)` to be rendered.

When you say that this approach will not work, do you have a concrete example of a text that fails to render here?

naveen521kk replied:
Pretty much on `Text("a" * 150)` text is still half-written by Cairo sadly.
![image](https://user-images.githubusercontent.com/49693820/94249133-147c4d80-ff3d-11ea-9998-02dd67a9742d.png)

behackl replied:
Where is this image from? When I compile the scene in this example:

```py
class Test(Scene):
    def construct(self):
        t = "a" * 150 + "b"
        text = Text(t, font='Arial', stroke_width=0)

        self.add(text.scale(0.3))
        self.wait(5)
```

I get

![image](https://user-images.githubusercontent.com/11851593/94250011-f143a500-ff20-11ea-9187-cbc78d3e580c.png)

The string is rendered and displayed in full here.

naveen521kk replied:
That image is from my browser when I opened the svg file.

naveen521kk replied:
Same script for me 
```py
class Test(Scene):
    def construct(self):
        t = "a" * 150 + "b"
        text = Text(t, font='Arial', stroke_width=0)

        self.add(text.scale(0.3))
        self.wait(5)
```
![image](https://user-images.githubusercontent.com/49693820/94254941-1e09b380-ff45-11ea-8f96-0c6117bb2959.png)

behackl replied:
I am clueless, I don't really see a reason for the different behavior you encounter. Did you compile this with an empty `media/texts` directory? (Just to exclude caching issues.)

The only significant difference is probably that you are using manim from Windows, right? (I've tried this fix on three different linux machines, I always get the full string rendered.)

naveen521kk replied:
Pretty much, yes, I was using the master branch. (Forgot to run poetry install). It works now.
![image](https://user-images.githubusercontent.com/49693820/94263144-3af8b380-ff52-11ea-90b9-7c84e9629c92.png)

leotrs commented on manim/mobject/svg/text_mobject.py:
```diff
@@ -297,7 +297,7 @@ def text2svg(self):
         if os.path.exists(file_name):
             return file_name
-        surface = cairo.SVGSurface(file_name, 600, 400)
+        surface = cairo.SVGSurface(file_name, len(self.text) * size, 400)
```
I hate hard-coding any part of this. Why was it set to 600, 400??? Oh man.

Question: can we set the surface size to cover the whole manim canvas? Would that slow things down considerably?

behackl replied:
While we could do that, that might not be enough: The example I have above would not fit on the entire canvas, but after scaling the text, it does.

Regarding render speed: I am not sure in how far the size of the canvas actually influences the required time; in the experiments I did I couldn't really observe an influence (although my change increases the canvas size on average by a factor of at least 2).

behackl commented:
As discussed on Discord (@huguesdevimeux, @naveen521kk): if the `size` is too large (e.g., `size=3000`, which is actually ridiculously large), then the same error as in #469 occurs.

I've tried to fix it in a similar way, but I fear that this hits some other limit than the canvas size: no matter how large I set the canvas, and no matter how I place the text, the error remains.

If someone else has an idea how to fix this, suggestions are welcome. Otherwise I see two possible ways forwards:

1. Simply ignore it, or
2. throw a `ValueError` when setting `size` larger than 1000 with a link to this PR.

huguesdevimeux commented:
If there is no fix, I think ignoring and warning the user that there is a problem is fine IMO

behackl commented:
> If there is no fix, I think ignoring and warning the user that there is a problem is fine IMO

@huguesdevimeux Good idea! I'll raise a warning if a font size larger than 500 or so is used.

Another issue: the `test_text.py` introduced here does not work: when I run `pytest` from the root directory of the repository, then this fails with
```
E       manim.mobject.svg.text_mobject.cairo.IOError: error while writing to output stream

manim/mobject/svg/text_mobject.py:300: cairo.IOError
=== short test summary info ===
FAILED tests/test_text.py::test_long_string - manim.mobject.svg.text_mobject.cairo.IOError: error while writing to output stream
```

The reason seems to be that the `media/texts` directory does not exist (at least when I create that manually, then the test passes). Do you think it would be a good idea to make this more robust in some sense? Maybe either by letting the call to `Text` create a directory if it does not exist, or otherwise reverting to using an appropriate temporary directory?

leotrs commented:
This LGTM but all tests are failing :(

behackl commented:
> This LGTM but all tests are failing :(

Yes, this is what I mentioned recently on Discord. `Text` and `Tex` need to be more robust regarding directory creation, I hope to submit a separate PR for this soon, this this here depends on that.

leotrs commented:
Ah I see. Well sounds like a great opportunity to use our handy new DEP bot.

leotrs commented:
@behackl This can now continue development, no?

Not trying to be pushy, just making sure :)

behackl commented:
I've merged master and applied the same solution/warning that I've added to `CairoText` to `PangoText` as well.

`PangoText` and `CairoText` share **a lot** of common code, if we intend to keep `CairoText` around for a bit longer, it might make sense to refactor these to classes a bit. The fact that I've added the solution to both in the very same way should not make an eventual refactor more complicated, but I'd like to not do it here immediately but rather observe how the situation with `PangoText` evolves.

leotrs commented:
The plan for Pango/CairoText is to start using the former instead of the latter (#609). Once that is merged, we will probably keep CairoText around for a bit until we are sure Pango is working well. Then, CairoText is going away completely.

So I wouldn't worry *too much* about keeping them aligned. As long as #609 is merged soon, we should be able to completely forget about CairoText soon as well.

leotrs commented:
@behackl don't stress about pango/cairo for this one ;)

also, would it be possible to add a test so we don't regress in the future?

behackl commented:
I've already added a test here to check that. I am not entirely sure, however, why the tests on this PR fail only for windows; I need to investigate a bit more.

behackl commented:
The situation here has not become any clearer, and I have absolutely no clue why changing the canvas size dynamically (in particular: making it _larger_ makes test fail on Windows.

Overall, the Text class is very messy. I propose to get back to this PR once we have managed to refactor Text a little.
